File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/components/UseSmileIDSampleSelectTrigger.kt
package com.usesmileid.sampleapps.ui.components

import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.smileid.designsystem.SmileDimens
import com.usesmileid.sampleapps.ui.theme.UseSmileIDSampleTheme

/** The design leads each trigger with an emoji rather than a glyph, so it carries no tint. */
@Composable
fun UseSmileIDSampleTriggerEmoji(emoji: String) = Text(
    text = emoji,
    style = UseSmileIDSampleTheme.type.inputFont.copy(fontSize = TRIGGER_EMOJI_SIZE),
)

/** Looks like an input, behaves like a button. Disabled is load-bearing: ID type stays greyed until a country is chosen. */
@Composable
fun UseSmileIDSampleSelectTrigger(
    value: String?,
    placeholder: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    testId: String? = null,
    leading: @Composable ((Color) -> Unit)? = null,
) {
    val colors = UseSmileIDSampleTheme.colors
    val contentColor = when {
        // The disabled pair, not textMuted on an almost-white surface, which does not read as disabled.
        !enabled -> colors.button.disabledText
        value != null -> colors.textTitle
        else -> colors.input.placeholder
    }
    Surface(
        onClick = onClick,
        enabled = enabled,
        modifier = modifier
            .fillMaxWidth()
            .semantics { role = Role.Button }
            .tagged(testId),
        shape = RoundedCornerShape(SmileDimens.radiusField),
        color = if (enabled) colors.input.background else colors.button.disabledBackground,
    ) {
        Row(
            modifier = Modifier
                .defaultMinSize(minHeight = SmileDimens.sizeControlMd)
                .border(
                    // The design outlines an actionable trigger in primary and a disabled one in border.
                    width = SmileDimens.borderWidthThin,
                    color = if (enabled) colors.primary else colors.input.border,
                    shape = RoundedCornerShape(SmileDimens.radiusField),
                )
                .padding(horizontal = SmileDimens.spacingMd, vertical = SmileDimens.spacingSm),
            horizontalArrangement = Arrangement.spacedBy(SmileDimens.spacingXs),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (leading != null) {
                // A minimum, not a fixed size: the leading slot may hold an emoji, which grows with font scale.
                Box(
                    modifier = Modifier.defaultMinSize(minWidth = SmileDimens.sizeIconMd, minHeight = SmileDimens.sizeIconMd),
                    contentAlignment = Alignment.Center,
                ) {
                    leading(contentColor)
                }
            }
            Text(
                // 15/600 in the design, where inputFont is 14/400.
                text = value ?: placeholder,
                style = UseSmileIDSampleTheme.type.inputFont.copy(
                    fontSize = TRIGGER_TEXT_SIZE,
                    fontWeight = FontWeight.SemiBold,
                ),
                color = contentColor,
                modifier = Modifier.weight(1f),
            )
            ChevronDownGlyph(tint = contentColor, size = CHEVRON_SIZE)
        }
    }
}

/** 15, 18 and 12 in the design; no token carries them. */
private val TRIGGER_TEXT_SIZE = 15.sp
private val TRIGGER_EMOJI_SIZE = 18.sp
private val CHEVRON_SIZE = 12.dp

File: flutter/sample_ui/lib/src/components/use_smileid_sample_select_trigger.dart
import 'package:flutter/material.dart';

import '../theme/use_smileid_sample_colors.dart';
import '../theme/use_smileid_sample_theme.dart';
import '../theme/use_smileid_sample_typography.dart';
import '../tokens/smile_tokens.dart';
import 'use_smileid_sample_glyphs.dart';

/// The design leads each trigger with an emoji rather than a glyph, so it carries no tint.
class UseSmileIDSampleTriggerEmoji extends StatelessWidget {
  /// Takes the emoji; the chosen country's flag, or a globe as placeholder.
  const UseSmileIDSampleTriggerEmoji({required this.emoji, super.key});

  /// The emoji to draw.
  final String emoji;

  @override
  Widget build(BuildContext context) => Text(
    emoji,
    style: UseSmileIDSampleType.inputFont.copyWith(fontSize: _emojiSize),
  );
}

/// Looks like an input, behaves like a button.
///
/// Disabled is load-bearing: the ID-type trigger stays greyed until a country is chosen.
class UseSmileIDSampleSelectTrigger extends StatelessWidget {
  /// A null [value] draws [placeholder] in the placeholder colour.
  const UseSmileIDSampleSelectTrigger({
    required this.value,
    required this.placeholder,
    required this.onTap,
    this.enabled = true,
    this.leading,
    this.testId,
    super.key,
  });

  /// The chosen value, null until one is chosen.
  final String? value;

  /// Shown while nothing is chosen.
  final String placeholder;

  /// What a tap does.
  final VoidCallback onTap;

  /// Whether the trigger can be tapped.
  final bool enabled;

  /// The leading emoji or glyph, handed the content colour.
  final Widget Function(Color tint)? leading;

  /// The `sample_*` id the screen supplies.
  final String? testId;

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(
      context,
    );
    // The disabled pair, not muted text on an almost-white surface, which does not read as disabled.
    final Color content = !enabled
        ? colors.button.disabledText
        : value != null
        ? colors.textTitle
        : colors.input.placeholder;
    return Semantics(
      identifier: testId,
      button: true,
      enabled: enabled,
      child: Material(
        color: enabled
            ? colors.input.background
            : colors.button.disabledBackground,
        shape: RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(SmileDimens.radiusField),
          // Outlined in PRIMARY when actionable, which is what says it can be tapped.
          side: BorderSide(
            color: enabled ? colors.primary : colors.input.border,
            width: SmileDimens.borderWidthThin,
          ),
        ),
        child: InkWell(
          onTap: enabled ? onTap : null,
          borderRadius: BorderRadius.circular(SmileDimens.radiusField),
          child: ConstrainedBox(
            constraints: const BoxConstraints(
              minHeight: SmileDimens.sizeControlMd,
            ),
            child: Padding(
              padding: const EdgeInsets.symmetric(
                horizontal: SmileDimens.spacingMd,
                vertical: SmileDimens.spacingSm,
              ),
              child: Row(
                children: <Widget>[
                  if (leading != null) ...<Widget>[
                    // A minimum, not a fixed box: an emoji grows with the font scale and clips.
                    ConstrainedBox(
                      constraints: const BoxConstraints(
                        minWidth: SmileDimens.sizeIconMd,
                        minHeight: SmileDimens.sizeIconMd,
                      ),
                      child: Center(child: leading!(content)),
                    ),
                    const SizedBox(width: SmileDimens.spacingXs),
                  ],
                  Expanded(
                    child: Text(
                      value ?? placeholder,
                      style: UseSmileIDSampleType.inputFont.copyWith(
                        fontSize: _triggerTextSize,
                        fontWeight: FontWeight.w600,
                        color: content,
                      ),
                    ),
                  ),
                  const SizedBox(width: SmileDimens.spacingXs),
                  UseSmileIDSampleGlyphs.chevronDown(
                    content,
                    size: _chevronSize,
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}

/// The trigger's label run.
const double _triggerTextSize = 15;

/// The leading emoji's run.
const double _emojiSize = 18;

/// The DOWN chevron; the list chevron is larger and points right.
const double _chevronSize = 12;
